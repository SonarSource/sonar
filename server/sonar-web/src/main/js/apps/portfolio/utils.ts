/*
 * SonarQube
 * Copyright (C) 2009-2016 SonarSource SA
 * mailto:contact AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
export function getNextRating(rating: number): number | undefined {
  return rating > 1 ? rating - 1 : undefined;
}

function getWorstSeverity(data: string): { severity: string; count: number } | undefined {
  const SEVERITY_ORDER = ['BLOCKER', 'CRITICAL', 'MAJOR', 'MINOR', 'INFO'];

  const severities: { [key: string]: number } = {};
  data.split(';').forEach(equality => {
    const [key, count] = equality.split('=');
    severities[key] = Number(count);
  });

  for (let i = 0; i < SEVERITY_ORDER.length; i++) {
    const count = severities[SEVERITY_ORDER[i]];
    if (count > 0) {
      return { severity: SEVERITY_ORDER[i], count };
    }
  }

  return undefined;
}

export function getEffortToNextRating(
  measures: Array<{ metric: { key: string }; value: string }>,
  metricKey: string
) {
  const measure = measures.find(measure => measure.metric.key === metricKey);
  if (!measure) {
    return undefined;
  }
  return getWorstSeverity(measure.value);
}
