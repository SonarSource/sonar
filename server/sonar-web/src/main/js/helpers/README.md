## Making Requests

Since 6.0 all API requests must provide CSRF protection token. In order to
help you to do so, we provide a set of useful helpers.

### Getting Started
Let's start with a simple GET request.

```
window.sonarqube.appStarted.then(function () {
  // some code here
  window.SonarRequest
      .getJSON(window.baseUrl + '/api/issues/search')
      .then(function (response) {
        // here 'response' contains the object representing the JSON output
      });
});
```

`window.sonarqube.appStarted` returns a `Promise` telling you when the
application is started and all the runtime libraries and hooks are installed.

`window.SonarRequest` contains all the helper methods to do an API requests.

`window.SonarRequest.getJSON` is a simplest helper to do an API call,
receive some data and parse it as JSON.

### API Documentation

#### `window.sonarqube.appStarted: Promise`
A promise telling when the application has started.
You must put your code in the `resolve` path.

#### `window.SonarRequest.request(url: string): Request`
Start making an API call. `url` must start from `window.baseUrl`.
Return a `Request` instance which has the following methods:
* `setMethod(method: string): Request` sets the http method, can be `GET`, `POST`, etc.
* `setData(data: object): Request` sets the request parameters`
* `submit(): Promise` sends the request

#### `window.SonarRequest.getJSON(url: string[, data: object]): Promise`
Send a GET request, get a response, parse it as JSON.

#### `window.SonarRequest.postJSON(url: string[, data: object]): Promise`
Send a POST request, get a response, parse it as JSON.

#### `window.SonarRequest.post(url: string[, data: object]): Promise`
Send a POST request, ignore the response content.

### Important to Know

Under the hood `Request` uses `fetch API`.

All `getJSON`, `postJSON` and `post` check the response status code.
They reject the Promise if the status is not in [200, 300).

### More Examples

#### Get the list of unresolved issues
```
window.SonarRequest.getJSON(
  window.baseUrl + '/api/issues/search',
  { resolved: false }
).then(function (response) {
  // response.issues contains the list of issues
});
```

#### Create new project
```
window.SonarRequest.post(
  window.baseUrl + '/api/projects/create',
  { key: 'sample', name: 'Sample' }
).then(function () {
  // the project has been created
});
```

#### Handle bad requests
```
window.SonarRequest.post(
  window.baseUrl + '/api/users/deactivate',
  { login: 'admin' }
).catch(function (error) {
  // error.response.status === 400
  // error.response.statusText === 'Bad Request'
  // To read the response:
  // error.response.json().then(function (jsonResponse) { ... });
});
```
