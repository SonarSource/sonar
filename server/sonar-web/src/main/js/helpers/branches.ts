/*
 * SonarQube
 * Copyright (C) 2009-2016 SonarSource SA
 * mailto:contact AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { sortBy } from 'lodash';
import { Branch, BranchType, ShortLivingBranch, LongLivingBranch } from '../app/types';

export const MAIN_BRANCH: Branch = {
  isMain: true,
  name: undefined,
  status: { qualityGateStatus: 'OK' },
  type: BranchType.LONG
};

const MAIN_BRANCH_DISPLAY_NAME = 'master';

export function isShortLivingBranch(branch: Branch | null): branch is ShortLivingBranch {
  return branch != null && !branch.isMain && branch.type === BranchType.SHORT;
}

export function getBranchDisplayName(branch: Branch): string {
  return branch.isMain ? MAIN_BRANCH_DISPLAY_NAME : branch.name;
}

export function getBranchName(branch: Branch): string | undefined {
  return branch.isMain ? undefined : branch.name;
}

export function sortBranchesAsTree(branches: Branch[]): Branch[] {
  const result: Branch[] = [];

  // main branch is always first
  const mainBranch = branches.find(branch => branch.isMain);
  if (mainBranch) {
    result.push(mainBranch);
  }

  // then process all branches where merge branch is empty
  getChildren().forEach(processBranch);

  return result;

  function getChildren(mergeBranch?: string): Branch[] {
    return sortBy(
      branches.filter(branch => !branch.isMain && branch.mergeBranch === mergeBranch),
      branch => !isShortLivingBranch(branch),
      'name'
    );
  }

  function processBranch(branch: ShortLivingBranch | LongLivingBranch) {
    result.push(branch);
    getChildren(branch.name).forEach(processBranch);
  }
}
