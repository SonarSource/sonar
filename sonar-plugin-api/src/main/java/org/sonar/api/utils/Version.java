/*
 * SonarQube
 * Copyright (C) 2009-2016 SonarSource SA
 * mailto:contact AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package org.sonar.api.utils;

import com.google.common.base.Splitter;
import java.util.List;
import javax.annotation.concurrent.Immutable;

import static java.lang.Integer.parseInt;
import static org.apache.commons.lang.StringUtils.substringAfter;
import static org.apache.commons.lang.StringUtils.substringBefore;
import static org.apache.commons.lang.StringUtils.trimToEmpty;

/**
 * TODO document equals/comparison without qualifier
 * @since 5.5
 */
@Immutable
public class Version implements Comparable<Version> {

  private static final String QUALIFIER_SEPARATOR = "-";
  private static final char SEQUENCE_SEPARATOR = '.';
  private static final Splitter SEQUENCE_SPLITTER = Splitter.on(SEQUENCE_SEPARATOR);

  private final int major;
  private final int minor;
  private final int patch;
  private final String qualifier;

  private Version(int major, int minor, int patch, String qualifier) {
    this.major = major;
    this.minor = minor;
    this.patch = patch;
    this.qualifier = qualifier;
  }

  public int major() {
    return major;
  }

  public int minor() {
    return minor;
  }

  public int patch() {
    return patch;
  }

  /**
   * @return non-null suffix
   */
  public String qualifier() {
    return qualifier;
  }

  public static Version parse(String text) {
    String s = trimToEmpty(text);
    String qualifier = substringAfter(s, QUALIFIER_SEPARATOR);
    if (!qualifier.isEmpty()) {
      s = substringBefore(s, QUALIFIER_SEPARATOR);
    }
    List<String> split = SEQUENCE_SPLITTER.splitToList(s);
    int major = 0;
    int minor = 0;
    int patch = 0;
    if (split.size() > 0) {
      major = parseSequence(split.get(0));
      if (split.size() > 1) {
        minor = parseSequence(split.get(1));
        if (split.size() > 2) {
          patch = parseSequence(split.get(2));
          if (split.size() > 3) {
            throw new IllegalArgumentException("Only 3 sequences are accepted");
          }
        }
      }
    }
    return new Version(major, minor, patch, qualifier);
  }

  public static Version create(int major, int minor) {
    return new Version(major, minor, 0, "");
  }

  public static Version create(int major, int minor, int patch) {
    return new Version(major, minor, patch, "");
  }

  public static Version create(int major, int minor, int patch, String qualifier) {
    return new Version(major, minor, patch, qualifier);
  }

  private static int parseSequence(String sequence) {
    if (sequence.isEmpty()) {
      return 0;
    }
    return parseInt(sequence);
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (!(o instanceof Version)) {
      return false;
    }
    Version other = (Version) o;
    return major == other.major && minor == other.minor && patch == other.patch;
  }

  @Override
  public int hashCode() {
    int result = major;
    result = 31 * result + minor;
    result = 31 * result + patch;
    return result;
  }

  @Override
  public int compareTo(Version other) {
    int c = major - other.major;
    if (c == 0) {
      c = minor - other.minor;
      if (c == 0) {
        c = patch - other.patch;
      }
    }
    return c;
  }

  @Override
  public String toString() {
    StringBuilder sb = new StringBuilder();
    sb.append(major).append(SEQUENCE_SEPARATOR).append(minor);
    if (patch > 0) {
      sb.append(SEQUENCE_SEPARATOR).append(patch);
    }
    if (!qualifier.isEmpty()) {
      sb.append(QUALIFIER_SEPARATOR).append(qualifier);
    }
    return sb.toString();
  }
}
