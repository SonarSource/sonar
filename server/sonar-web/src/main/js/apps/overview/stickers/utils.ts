/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { getHostUrl } from '../../../helpers/urls';

export enum StickerType {
  badge = 'badge',
  card = 'card',
  marketing = 'marketing'
}

export interface StickerOptions {
  color?: 'white' | 'black' | 'orange';
}

export function getStickerUrl(type: StickerType, options: StickerOptions) {
  switch (type) {
    case StickerType.marketing:
      return `${getHostUrl()}/images/stickers/sonarcloud-${options.color || 'white'}.svg`;
    case StickerType.card:
      return '';
    case StickerType.badge:
      return '';
  }
}
