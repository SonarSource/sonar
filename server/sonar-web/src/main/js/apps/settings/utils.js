/*
 * SonarQube
 * Copyright (C) 2009-2016 SonarSource SA
 * mailto:contact AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { translate, hasMessage } from '../../helpers/l10n';
import { TYPE_PROPERTY_SET, TYPE_BOOLEAN, TYPE_SINGLE_SELECT_LIST } from './constants';

export const DEFAULT_CATEGORY = 'general';

export function getPropertyName (definition) {
  const key = `property.${definition.key}.name`;
  return hasMessage(key) ? translate(key) : definition.name;
}

export function getPropertyDescription (definition) {
  const key = `property.${definition.key}.description`;
  return hasMessage(key) ? translate(key) : definition.description;
}

export function getCategoryName (category) {
  const key = `property.category.${category}`;
  return hasMessage(key) ? translate(key) : category;
}

export function getSubCategoryName (category, subCategory) {
  const key = `property.category.${category}.${subCategory}`;
  return hasMessage(key) ? translate(key) : getCategoryName(subCategory);
}

export function getSubCategoryDescription (category, subCategory) {
  const key = `property.category.${category}.${subCategory}.description`;
  return hasMessage(key) ? translate(key) : null;
}

export function getUniqueName (definition, index = null) {
  const indexSuffix = index != null ? `[${index}]` : '';
  return `settings[${definition.key}]${indexSuffix}`;
}

export function getSettingValue (setting) {
  if (setting.definition.multiValues) {
    return setting.values;
  } else if (setting.definition.type === TYPE_PROPERTY_SET) {
    return setting.fieldsValues;
  } else {
    return setting.value;
  }
}

export function isEmptyValue (definition, value) {
  if (value == null) {
    return true;
  } else if (definition.type === TYPE_BOOLEAN) {
    return false;
  } else {
    return value.length === 0;
  }
}

export function getEmptyValue (definition) {
  if (definition.type === TYPE_BOOLEAN || definition.type === TYPE_SINGLE_SELECT_LIST) {
    return null;
  } else {
    return '';
  }
}

export function isDefaultOrInherited (setting) {
  return !!setting.default || !!setting.inherited;
}

/**
 * Check if provided definition is a complex one
 * Complex definitions are property sets or multi values
 * @param definition
 * @returns {boolean}
 */
export function isComplexDefinition (definition) {
  return definition.type === TYPE_PROPERTY_SET || definition.multiValues === true;
}

/**
 * Get and format the default value
 * @param definition
 * @returns {string}
 */
export function getDefaultValue (definition) {
  if (definition.defaultValue == null) {
    return '<no value>';
  } else if (definition.type === TYPE_BOOLEAN) {
    return definition.defaultValue ? translate('settings.boolean.true') : translate('settings.boolean.false');
  } else {
    return definition.defaultValue;
  }
}
