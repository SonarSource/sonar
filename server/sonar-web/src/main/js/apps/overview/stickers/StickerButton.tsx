/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import * as classNames from 'classnames';
import { StickerType } from './utils';
import { translate } from '../../../helpers/l10n';

interface Props {
  onClick: (type: StickerType) => void;
  selected: boolean;
  type: StickerType;
  url: string;
}

export default class StickerButton extends React.PureComponent<Props> {
  handleClick = (event: React.SyntheticEvent<HTMLAnchorElement>) => {
    event.preventDefault();
    this.props.onClick(this.props.type);
  };
  render() {
    return (
      <a
        className={classNames('sticker-button', { selected: this.props.selected })}
        href=""
        onClick={this.handleClick}>
        <img src={this.props.url} alt={translate('overview.stickers', this.props.type, 'alt')} />
      </a>
    );
  }
}
