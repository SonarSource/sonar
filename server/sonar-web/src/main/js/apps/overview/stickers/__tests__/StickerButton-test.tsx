/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { shallow } from 'enzyme';
import StickerButton from '../StickerButton';
import { StickerType } from '../utils';
import { click } from '../../../../helpers/testUtils';

it('should display correctly', () => {
  expect(getWrapper()).toMatchSnapshot();
  expect(getWrapper({ selected: true })).toMatchSnapshot();
});

it('should return the sticker type on click', () => {
  const onClick = jest.fn();
  const wrapper = getWrapper({ onClick });
  click(wrapper.find('a'));
  expect(onClick).toHaveBeenCalledWith(StickerType.marketing);
});

function getWrapper(props = {}) {
  return shallow(
    <StickerButton
      onClick={jest.fn()}
      selected={false}
      type={StickerType.marketing}
      url="http://foo.bar"
      {...props}
    />
  );
}
