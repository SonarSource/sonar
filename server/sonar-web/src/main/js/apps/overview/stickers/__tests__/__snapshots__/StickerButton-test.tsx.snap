// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display correctly 1`] = `
<a
  className="sticker-button"
  href=""
  onClick={[Function]}
>
  <img
    alt="overview.stickers.marketing.alt"
    src="http://foo.bar"
  />
</a>
`;

exports[`should display correctly 2`] = `
<a
  className="sticker-button selected"
  href=""
  onClick={[Function]}
>
  <img
    alt="overview.stickers.marketing.alt"
    src="http://foo.bar"
  />
</a>
`;
